### PR TITLE
trigger instance refresh on launch_template change

### LINKED ISSUE #25259: Run an ASG instance refresh on TF configuration change
### What would you like Teleport to do?

As a user editing the Teleport ASG with Terraform (updating config or Teleport version) I want the changes to apply immediately.

### What problem does this solve?

Changes in configuration don't trigger an instance refresh, thus are not applied unless machines are deleted/restarted.
This causes a lot of confusion and time lost troubleshooting why the configuration change did not apply. This can also cause security issues when users think they have updated to a version containing a security fix while still running vulnerable images.

### If a workaround exists, please include it.

Trigger instance rollout manually.

### Implementation details

See https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/autoscaling_group#automatically-refresh-all-instances-after-the-group-is-updated

### PR DESCRIPTION
**Pull request by GavinFrazar**
changelog: The "ha-autoscale-cluster" terraform module now support default AWS resource tags and ASG instance refresh on configuration or launch template changes.

Closes https://github.com/gravitational/teleport/issues/25259

launch template changes will trigger a refresh by default, *unless* you use `version = "$Latest"` instead of referring to the template latest version explicitly:

> A refresh will not start when version = "$Latest" is configured in the launch_template block. To trigger the instance refresh when a launch template is changed, configure version to use the latest_version attribute of the aws_launch_template resource.

### CODE REVIEW

**Review by hugoShaka — APPROVED** [review]
LGTM if this change was tested on a real `ha-autoscale-cluster` deployment?

**Comment by GavinFrazar** [comment]
> LGTM if this change was tested on a real `ha-autoscale-cluster` deployment?

I couldn't get the module to work before because of our tagging policies, but I've since tested this change.

I'm glad I did because this change alone isn't enough - I thought that `instance_refresh` was the default behavior as long as the template version was specified instead of `$Latest`, but I had to add an `instance_refresh` section to the ASG in addition to my changes here to get it to auto refresh on config change.

I'll push an updated PR that includes my default tag propagation changes as well, since that was pretty painful to figure out.

File: examples/aws/terraform/ha-autoscale-cluster/auth_asg.tf [hunk 1]
```diff
@@ -33,6 +33,26 @@ resource "aws_autoscaling_group" "auth" {
...
+  dynamic "instance_refresh" {
+    for_each = var.enable_auth_asg_instance_refresh ? [1] : []
+    content {
+      strategy = "Rolling"
+      preferences {
+        auto_rollback          = true
+        min_healthy_percentage = 50
+      }
+    }
+  }
```
Comment (hugoShaka) [inline examples/aws/terraform/ha-autoscale-cluster/auth_asg.tf:54]: We don't technically support rolling restart for Auth. For proxy it's OK, but Auth should all be restarted at the same time during updates. I don't think there's an easy way to do it, maybe `min_healthy_percentage = 0`.

Also we really don't want to automatically rollback auth as not all updates can be rolled back. You might have to rollback the state at the same time.

Comment (GavinFrazar) [reply examples/aws/terraform/ha-autoscale-cluster/auth_asg.tf:54]: This is a good point, but we also tell customers to scale auth servers to 1 when they upgrade or downgrade Teleport right?

Is there any problem with a rolling instance refresh when configuration *other than teleport version* is changed?

**Comment by webvictim** [comment]
The instance refresh shouldn't be enabled by default IMO - it's a breaking change. We don't version the Terraform code so anyone doing a `git pull` needs a consistent experience which won't suddenly cause their instances to restart.

**Comment by GavinFrazar** [comment]
@hugoShaka @webvictim 

Are you both good with this:
1. disable the auto-rollback for auth servers' ASG
2. keep the instance refresh code, but default to `false` instead of `true` on the input vars

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 2]
```diff
@@ -240,3 +240,30 @@ variable "teleport_auth_type" {
...
+variable "default_tags" {
+  type    = map(string)
+  default = {}
+}
+
+// Whether to trigger instance refresh rollout for Teleport Auth servers when
+// the launch template or resource tags change.
+variable "enable_auth_asg_instance_refresh" {
+  type    = bool
+  default = true
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:254]:
```suggestion
  default = false
```

As discussed we shouldn't automatically kick off instance refreshes as it'd be a breaking change for folks currently using the module who schedule their refreshes manually during maintenance windows. An unfortunate fact of not having versioned this module :(

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 3]
```diff
@@ -240,3 +240,30 @@ variable "teleport_auth_type" {
...
+}
+
+// Whether to trigger instance refresh rollout for Teleport Proxy servers when
+// the launch template or resource tags change.
+variable "enable_proxy_asg_instance_refresh" {
+  type    = bool
+  default = true
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:261]:
```suggestion
  default = false
```

File: examples/aws/terraform/ha-autoscale-cluster/vars.tf [hunk 4]
```diff
@@ -240,3 +240,30 @@ variable "teleport_auth_type" {
...
+}
+
+// Whether to trigger instance refresh rollout for Teleport Node servers when
+// the launch template or resource tags change.
+variable "enable_node_asg_instance_refresh" {
+  type    = bool
+  default = true
```
Comment (webvictim) [inline examples/aws/terraform/ha-autoscale-cluster/vars.tf:268]:
```suggestion
  default = false
```

**Review by webvictim — COMMENTED** [review]
Thanks for adding this! I think having a way to automatically trigger ASG refresh if people want it is great, but we shouldn't enable it by default.

Could you also please:
- expose the new variables in https://github.com/gravitational/teleport/blob/master/examples/aws/terraform/ha-autoscale-cluster/Makefile (as is done with others)
- update https://github.com/gravitational/teleport/blob/master/docs/pages/admin-guides/deploy-a-cluster/deployments/aws-ha-autoscale-cluster-terraform.mdx to add the new variables

**Comment by GavinFrazar** [comment]
> Thanks for adding this! I think having a way to automatically trigger ASG refresh if people want it is great, but we shouldn't enable it by default.
> 
> Could you also please:
> 
> * expose the new variables in https://github.com/gravitational/teleport/blob/master/examples/aws/terraform/ha-autoscale-cluster/Makefile (as is done with others)
> * update https://github.com/gravitational/teleport/blob/master/docs/pages/admin-guides/deploy-a-cluster/deployments/aws-ha-autoscale-cluster-terraform.mdx to add the new variables

done in [ede89dd](https://github.com/gravitational/teleport/pull/46324/commits/ede89dd406385e6b0c2519bcad0020ad22e586b5)

**Review by webvictim — APPROVED** [review]

**Comment by webvictim** [comment]
I think you can get away with only backporting this to v16 if you want - we don't version the module, and most folks either pin a specific commit hash or just use the latest version from `master`. The v16 backport would only really be needed make the changelog entry appear along with 16.4.1 or whatever release is up next. It's up to you though.

**Review by greedy52 — APPROVED** [review]

**Pull request merged**